### PR TITLE
Rebuild sharing on top of Runtime-Storage abstraction

### PR DESCRIPTION
**Pull request by sjmiles**
Other things:
- collecting some file diaspora into shell/lib
- removing legacy implementations

### CODE REVIEW

File: shell/app-shell/app-shell.js [hunk 1]
```diff
@@ -73,15 +83,15 @@ const template = html`
     on-search="_onStateData"
   ></arc-planner>
 
-  <shell-stores
+  <!-- <shell-stores
```
Comment (shaper) [inline shell/app-shell/app-shell.js]: Is this intentionally commented out?

File: artifacts/Restaurants/Restaurants.recipes [hunk 2]
```diff
@@ -20,10 +20,10 @@ recipe &displayRestaurants
     list <- list
 
 recipe
-  use as person
+  map 'SYSTEM_user' as user
```
Comment (shaper) [inline artifacts/Restaurants/Restaurants.recipes]: Is this the new way to get the system user?

File: shell/app-shell/app-shell.js [hunk 3]
```diff
@@ -246,6 +257,16 @@ class AppShell extends Xen.Debug(Xen.Base, log) {
       state.showhint = Boolean(state.suggestions && state.suggestions.length > 0);
     }
   }
+  _updateSerialization(state) {
+    // TODO(sjmiles): wait 4s for context :(
```
Comment (shaper) [inline shell/app-shell/app-shell.js]: What is the longer term approach here, should we file an issue to discuss with @shans?

File: shell/app-shell/elements/arc-host.js [hunk 4]
```diff
@@ -50,7 +51,20 @@ class ArcHost extends Xen.Debug(Xen.Base, log) {
     // empty serialization which is ''
     if (id && context && pendingSerialization != null) {
       state.pendingSerialization = null;
+      // TODO(sjmiles): if we `_setState` we trigger invalidation if the return
```
Comment (shaper) [inline shell/app-shell/elements/arc-host.js]: This looks like the answer to my question above, I just want to make sure we talk about doing this soon.

File: shell/app-shell/elements/cloud-data.js [hunk 5]
```diff
@@ -36,13 +38,12 @@ const template = html`
...
-    config="{{config}}"
-    userid="{{userid}}"
+  <user-geolocation on-coords="_onCoords"></user-geolocation>
+  <user-context
     context="{{context}}"
-    arc="{{arc}}"
-    on-friends="_onForward"
-  ></fb-user-context>
+    userid="{{userid}}"
+    coords="{{coords}}"
```
Comment (shaper) [inline shell/app-shell/elements/cloud-data.js]: Just to make sure you intended it, this PR looks like an omnibus one that also incorporates rework of user coordinates and some description fixes.

File: shell/app-shell/elements/cloud-data.js [hunk 6]
```diff
@@ -58,7 +59,7 @@ const template = html`
     on-serialization="_onForward"
   ></cloud-arc>
 
-  <cloud-steps
+  <!-- <cloud-steps
```
Comment (shaper) [inline shell/app-shell/elements/cloud-data.js]: Is this commented out chunk intentional?

File: shell/app-shell/elements/sharing/single-user-context.js [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+      }
+    };
+  }
+  _onShareChanged(share) {
+    // TODO(sjmiles): schedule field processing to avoid re-entrancy
+    //log('SingleUserContext:onShareChanged', share);
+    const arcid = share.parent.key;
+    const userid = share.parent.path.split('/')[2];
+    const meta = share.data.metadata || Object;
+    if (share.disposed || meta.share < 2) {
```
Comment (shaper) [inline shell/app-shell/elements/sharing/single-user-context.js]: Didn't we have constants/enums for the `2` and `1`, maybe using those would be helpful

File: shell/app-shell/elements/sharing/single-user-context.js [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+          //  (2) looks like there are double `remove` events in the queue. Bug?
+          // In general, to avoid interleaving we'll probably need to
+          // use a stack to process changes async to receiving them.
+          // Parallel processing works as of now ... feature?
+          store.remove(entity.id);
+        }
+      });
+    }
+    else {
+      const uid = store.id.split('|').slice(-2, -1).pop();
```
Comment (shaper) [inline shell/app-shell/elements/sharing/single-user-context.js]: Didn't we add something to runtime to allow getting uid in some other manner?

File: shell/app-shell/elements/shell-stores.js [hunk 9]
```diff
@@ -9,17 +9,17 @@ subject to an additional IP rights grant found at http://polymer.github.io/PATEN
...
 import Xen from '../../components/xen/xen.js';
-import ArcsUtils from '../lib/arcs-utils.js';
+import ArcsUtils from '../../lib/arcs-utils.js';
 import './arc-store.js';
 import {FbStore} from './fb-data/FbStore.js';
 
 // templates
 const template = Xen.html`
-  <arc-store arc="{{arc}}" data="{{themeData}}" options="{{themeStoreOptions}}" on-change="_onShellThemeChange"></arc-store>
+  <!-- <arc-store arc="{{arc}}" data="{{themeData}}" options="{{themeStoreOptions}}" on-change="_onShellThemeChange"></arc-store>
```
Comment (shaper) [inline shell/app-shell/elements/shell-stores.js]: Here and below in this file, are these commented out intentionally?

File: shell/app-shell/elements/shell-stores.js [hunk 10]
```diff
@@ -9,17 +9,17 @@ subject to an additional IP rights grant found at http://polymer.github.io/PATEN
...
   <arc-store arc="{{arc}}" data="{{userStoreData}}" options="{{userStoreOptions}}"></arc-store>
-  <arc-store arc="{{arc}}" data="{{usersStoreData}}" options="{{usersStoreOptions}}"></arc-store>
+  <arc-store arc="{{arc}}" data="{{usersStoreData}}" options="{{usersStoreOptions}}"></arc-store>-->
```
Comment (shaper) [inline shell/app-shell/elements/shell-stores.js]: What's the difference between `userStoreData` and `usersStoreData`? Not needed for this PR but naming so it's clearer could be useful

File: shell/source/ArcsLib.js [hunk 11]
```diff
@@ -30,15 +32,18 @@ import {KeyManager} from '../../runtime/ts-build/keymgmt/manager.js';
 //Tracing.enable();
 
 const Arcs = {
-  version: '0.3',
+  version: '0.4',
```
Comment (shaper) [inline shell/source/ArcsLib.js]: I thought we were already on 0.4, maybe this is from an older branch? Also the Firebase changes below and above, are those ok, should they be part of this PR?

File: shell/test/specs/demo-tests.js [hunk 12]
```diff
@@ -463,7 +463,7 @@ describe('Arcs demos', function() {
     // });
   });
 
-  it('can buy gifts', /** @this Context */ function() {
+  it.skip('can buy gifts', /** @this Context */ function() {
```
Comment (shaper) [inline shell/test/specs/demo-tests.js]: Is it ok to just skip these two tests or should we file issues?

File: shell/app-shell/app-shell.js (hunk 1 above)
Comment (sjmiles) [reply shell/app-shell/app-shell.js]: Yes, shell-stores is deprecated. I'll wait a bit until I'm confident I don't need it, then remove it completely.

File: artifacts/Restaurants/Restaurants.recipes (hunk 2 above)
Comment (sjmiles) [reply artifacts/Restaurants/Restaurants.recipes]: It's where I landed, but we can tweak it as needed.

File: shell/app-shell/elements/cloud-data.js (hunk 5 above)
Comment (sjmiles) [reply shell/app-shell/elements/cloud-data.js]: There are some nuances. I had to do some modifications on User (and Geolocation) in order to make Restaurant demo function under this PR, I didn't want to leave it broken.

File: shell/app-shell/elements/shell-stores.js (hunk 10 above)
Comment (sjmiles) [reply shell/app-shell/elements/shell-stores.js]: shell-stores is deprecated; but `userStoreData` is current user data, `usersStoreData` is the list of all users.

File: shell/app-shell/elements/shell-stores.js (hunk 9 above)
Comment (sjmiles) [reply shell/app-shell/elements/shell-stores.js]: shell-stores itself is deprecated. These comments are vestiges of me deprecating it piece-meal until I was finally able to detach the whole thing. It does look sloppy. I guess I don't like deleting things until after some grace-period; maybe this is a bad habit.

File: shell/test/specs/demo-tests.js (hunk 12 above)
Comment (sjmiles) [reply shell/test/specs/demo-tests.js]: Both :) ... will open discussion.

File: shell/source/ArcsLib.js (hunk 11 above)
Comment (sjmiles) [reply shell/source/ArcsLib.js]: This version number isn't used anywhere and it had just become stale. I noticed and updated it. Technically not necessary for this PR but I hate to lose little changes like this, so I tend to leave them in.

Removing shell's access to Firebase in favor of Storage abstractions is part of this effort. Firebase marshaling in runtime has changed recently and it needed some cleanup here. E.g. shell's `firebase.js` is moved to `shell/lib` in this PR. There's more things to be done, I did try to keep it to a minimum. :)

File: shell/app-shell/app-shell.js (hunk 3 above)
Comment (sjmiles) [reply shell/app-shell/app-shell.js]: Ideally the context logic sends a signal when context meets some heuristic for stable. I temporarily just skipped that work in favor of the timeout.

**Pull request merged**